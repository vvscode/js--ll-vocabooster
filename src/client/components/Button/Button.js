import React from 'react';
import PropTypes from 'prop-types';
import './Button.less';

class BaseButton extends React.Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node,
    ]),
    type: PropTypes.string,
    onClick: PropTypes.func.isRequired,
    style: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    disabled: PropTypes.bool,
  };

  static defaultProps = {
    children: null,
    type: 'default',
    style: {},
    disabled: false,
  };

  render() {
    const { type, children, onClick, style, disabled } = this.props;
    let className = '';

    switch (type) {
      case 'default':
        className = `Button__button Button__button-transparent`;
        break;
      case 'primary':
        className = 'Button__button';
        break;
      default:
        className = 'Button__button';
    }

    return (
      <button
        className={className}
        style={style}
        onClick={onClick}
        disabled={disabled}
      >
        {children}
      </button>
    );
  }
}

export default BaseButton;
