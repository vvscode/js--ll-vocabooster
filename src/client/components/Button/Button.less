@import '@client/styles/variables.css';

.Button {
  &__button {
    border-radius: 5px;
    height: 40px;
    line-height: 40px;
    padding: 0 20px;
    background-color: var(--purple);
    font-size: 14px;
    font-weight: 600;
    color: white;
    box-sizing: content-box;
    transition: all 0.5s;
    border: 1px solid var(--purple);
    cursor: pointer;
    text-align: center;

    &:hover {
      color: var(--purple);
      background: white;
    }

    &:disabled {
      background-color: var(--background-child);
      color: var(--greyish);
      border-color: var(--greyish);
    }
  }

  &__button-transparent {
    background: transparent;
    color: var(--purple);
  }
}
