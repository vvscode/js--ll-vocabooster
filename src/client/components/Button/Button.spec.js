import React from 'react';
import { shallow } from 'enzyme';
import Button from './Button';

describe('Button', () => {
  let props;
  let button;

  beforeEach(() => {
    props = {
      /* List all properties with 'undefined' values */
      onClick: jest.fn(),
      style: {},
    };

    button = () => shallow(<Button {...props} />);
  });

  it('should user correct className for each type', () => {
    props.type = 'default';
    expect(button()).toMatchSnapshot('explicit default type');

    props.type = 'primary';
    expect(button()).toMatchSnapshot('primary type');

    props.type = 'xxx';
    expect(button()).toMatchSnapshot('implicit default type');
  });

  it('should use `disabled = true` property correctly', () => {
    props.disabled = true;
    const btn = button();
    expect(btn).toMatchSnapshot('disabled button');
  });

  it('should use `disabled = false` property correctly', () => {
    props.disabled = false;
    const btn = button();
    expect(btn).toMatchSnapshot('enabled button');
  });
  it('renders children correctly', () => {
    props.children = <div>Test</div>;
    expect(button).toMatchSnapshot('enabled button');
  });
});
