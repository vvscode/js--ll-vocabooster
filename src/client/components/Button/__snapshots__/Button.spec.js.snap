// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders children correctly: enabled button 1`] = `[Function]`;

exports[`Button should use \`disabled = false\` property correctly: enabled button 1`] = `
<button
  className="Button__button Button__button-transparent"
  disabled={false}
  onClick={[MockFunction]}
  style={Object {}}
/>
`;

exports[`Button should use \`disabled = true\` property correctly: disabled button 1`] = `
<button
  className="Button__button Button__button-transparent"
  disabled={true}
  onClick={[MockFunction]}
  style={Object {}}
/>
`;

exports[`Button should user correct className for each type: explicit default type 1`] = `
<button
  className="Button__button Button__button-transparent"
  disabled={false}
  onClick={[MockFunction]}
  style={Object {}}
/>
`;

exports[`Button should user correct className for each type: implicit default type 1`] = `
<button
  className="Button__button"
  disabled={false}
  onClick={[MockFunction]}
  style={Object {}}
/>
`;

exports[`Button should user correct className for each type: primary type 1`] = `
<button
  className="Button__button"
  disabled={false}
  onClick={[MockFunction]}
  style={Object {}}
/>
`;
